src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs

```diff
@@ -809,31 +809,55 @@ private static LiftingResult UserDefinedBinaryOperatorCanBeLifted(TypeSymbol lef
             // SPEC: better than all other function members, then the function member invocation is ambiguous and a binding-time 
             // SPEC: error occurs.
 
-            // UNDONE: This is a naive quadratic algorithm; there is a linear algorithm that works. Consider using it.
+            int i;
             var candidates = result.Results;
-            for (int i = 1; i < candidates.Count; ++i)
+            for (i = 0; i < candidates.Count; i++)
             {
-                if (candidates[i].Kind != OperatorAnalysisResultKind.Applicable)
+                if (candidates[i].Kind == OperatorAnalysisResultKind.Applicable)
+                {
+                    break;
+                }
+            }
+
+            for (int j = i + 1; j < candidates.Count; j++)
+            {
+                if (candidates[j].Kind == OperatorAnalysisResultKind.Inapplicable)
                 {
                     continue;
                 }
 
-                // Is this applicable operator better than every other applicable method?
-                for (int j = 0; j < i; ++j)
+                var better = BetterOperator(candidates[i].Signature, candidates[j].Signature, left, right, ref useSiteDiagnostics);
+                if (better == BetterResult.Left)
                 {
-                    if (candidates[j].Kind == OperatorAnalysisResultKind.Inapplicable)
-                    {
-                        continue;
-                    }
-                    var better = BetterOperator(candidates[i].Signature, candidates[j].Signature, left, right, ref useSiteDiagnostics);
-                    if (better == BetterResult.Left)
-                    {
-                        candidates[j] = candidates[j].Worse();
-                    }
-                    else if (better == BetterResult.Right)
-                    {
-                        candidates[i] = candidates[i].Worse();
-                    }
+                    candidates[j] = candidates[j].Worse();
+                }
+                else if (better == BetterResult.Right)
+                {
+                    candidates[i] = candidates[i].Worse();
+                    i = j;
+                }
+            }
+
+            for (int j = 0; j < candidates.Count; j++)
+            {
+                if (j == i || candidates[j].Kind == OperatorAnalysisResultKind.Inapplicable)
+                {
+                    continue;
+                }
+
+                var better = BetterOperator(candidates[i].Signature, candidates[j].Signature, left, right, ref useSiteDiagnostics);
+                if (better == BetterResult.Left)
+                {
+                    candidates[j] = candidates[j].Worse();
+                }
+                else if (better == BetterResult.Right)
+                {
+                    candidates[i] = candidates[i].Worse();
+                    break;
+                }
+                else
+                {
+                    break;
```

Comment:
This code appears to correctly modify the array so that there is a single applicable member when overload resolution succeeds. However, when there is more than one most specific candidate, this code may leave in the array some members marked as applicable even though some other member may be more specific (I think). We do have a way of seeing the candidate set through a public API in `SemanticModel` (I think). So it appears we may yield the wrong result in that particular case.

In the case that there is more than one most specific candidate (which will result in an error), I don't mind having the quadratic loop to precisely limit the set of candidates and eliminate those for which there is some more specific candidate. We could still have the code be fast in the non-error case, which is by far more common.

All of this is theory. I would ask @alrz to please either tell my why my theory is wrong, or craft a test that confirms my suspicion (by having a set of user-defined overloaded operators in a class hierarchy, an ambiguous use of the operator, and using `SemanticModel` to see the not-quite-correct overload set). And if I was right, I think the code should be adjusted, in the more-than-one-candidate case, to eliminate candidates for which some other method is more specific.